### PR TITLE
Track creation locations.

### PR DESCRIPTION
**Pull request by jacob314**
This CL is fine to land before the corresponding changes in Flutter or
Flutter Engine as without those changes source locations just show
up as missing. The source location menu item shows up as disabled
for widgets or render objects without a source location.
Note that RenderObjects will not have source locations even after the
Flutter change as there isn't a clear benefit in tracking them.

### CODE REVIEW

File: src/io/flutter/view/InspectorPanel.java [hunk 1]
```diff
@@ -163,6 +168,14 @@ private InspectorService getInspectorService() {
   }
 
   void setActivate(boolean enabled) {
+    // TODO(jacobr): where is the right place to be caching the pub roots.
```
Comment (jacob314) [inline src/io/flutter/view/InspectorPanel.java]: I'd like to cache the pubRoots somewhere as I do use the pub roots every  time a node is styled to determine whether it should be bold or not. If there is an event I can listen for when pub roots change that is also reasonable. I assume pub roots can't change on a hot reload.

File: src/io/flutter/inspector/DiagnosticsNode.java [hunk 2]
```diff
@@ -358,6 +358,17 @@ boolean hasException() {
...
 
+  public boolean hasCreationLocation() {
+    return json.has("creationLocation");
+  }
+
+  public Location getCreationLocation() {
+    if (!hasCreationLocation()) {
+      return null;
+    }
+    return new Location(json.getAsJsonObject("creationLocation"), null);
```
Comment (devoncarew) [inline src/io/flutter/inspector/DiagnosticsNode.java]: Not sure how frequently this is called; consider caching the `Location` instance in a private field?

File: src/io/flutter/inspector/Location.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import com.intellij.xdebugger.impl.XSourcePositionImpl;
+import io.flutter.utils.JsonUtils;
+
+import java.util.ArrayList;
+
+public class Location {
+  final JsonObject json;
+  Location parent;
+
+  public Location(JsonObject json, Location parent) {
```
Comment (devoncarew) [inline src/io/flutter/inspector/Location.java]: we're not using the `parent` param here?

File: src/io/flutter/inspector/Location.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.xdebugger.XSourcePosition;
+import com.intellij.xdebugger.impl.XSourcePositionImpl;
+import io.flutter.utils.JsonUtils;
+
+import java.util.ArrayList;
+
+public class Location {
+  final JsonObject json;
+  Location parent;
```
Comment (devoncarew) [inline src/io/flutter/inspector/Location.java]: `parent` is never assigned to

File: src/io/flutter/view/InspectorPanel.java (hunk 1 above)
Comment (devoncarew) [reply src/io/flutter/view/InspectorPanel.java]: Perhaps in the `FlutterApp`? Or at least something that's per-debug session. I think it's reasonable to assume that they won't change during a debug session.

Comment (devoncarew) [reply src/io/flutter/view/InspectorPanel.java]: What happens if we have two sessions running at the same time? Does the inspector become unhappy?

You can test this by launching an app against one device and then a 2nd.

Comment (jacob314) [reply src/io/flutter/view/InspectorPanel.java]: Cached on the FlutterApp for now. Good point about having two session running at the same time. We definitely don't handle that case currently.

File: src/io/flutter/inspector/DiagnosticsNode.java (hunk 2 above)
Comment (jacob314) [reply src/io/flutter/inspector/DiagnosticsNode.java]: Went ahead and cached the Location object.

File: src/io/flutter/inspector/Location.java (hunk 3 above)
Comment (jacob314) [reply src/io/flutter/inspector/Location.java]: the usage isn't obvious but we are on line 29. the behavior is the file from the parent is used if a fileName is not specified as is the case for parameter locations.
Notice that on line 71 we pass in parent as "this".

**Comment by jacob314** [comment]
@devoncarew  you are right that we handle multiple sessions at the same time badly. I'll send a follow up Cl that handles that in general. Based on offline discussion, the simplest solution is to ensure we only show the inspector for the first running session.

**Pull request merged**